packages/babel-plugin-transform-for-of/src/index.js

```diff
@@ -189,156 +160,74 @@ export default declare((api, options) => {
           right.isGenericType("Array") ||
           t.isArrayTypeAnnotation(right.getTypeAnnotation())
         ) {
-          replaceWithArray(path);
+          path.replaceWith(_ForOfStatementArray(path));
           return;
         }
 
-        const { node } = path;
-        const build = pushComputedProps(path, state);
-        const declar = build.declar;
-        const loop = build.loop;
-        const block = loop.body;
+        if (!state.availableHelper(builder.helper)) {
+          // Babel <7.9.0 doesn't support this helper
+          transformWithoutHelper(loose, path, state);
+          return;
+        }
+
+        const { node, parent, scope } = path;
+        const left = node.left;
+        let declar;
+
+        const stepKey = scope.generateUid("step");
+        const stepValue = t.memberExpression(
+          t.identifier(stepKey),
+          t.identifier("value"),
+        );
+
+        if (
+          t.isIdentifier(left) ||
+          t.isPattern(left) ||
+          t.isMemberExpression(left)
+        ) {
+          // for (i of test), for ({ i } of test)
+          declar = t.expressionStatement(
+            t.assignmentExpression("=", left, stepValue),
+          );
+        } else if (t.isVariableDeclaration(left)) {
+          // for (let i of test)
+          declar = t.variableDeclaration(left.kind, [
+            t.variableDeclarator(left.declarations[0].id, stepValue),
+          ]);
+        } else {
+          throw path
+            .get("left")
+            .buildCodeFrameError(
+              `Unknown node type ${left.type} in ForStatement`,
+            );
+        }
 
         // ensure that it's a block so we can take all its statements
         path.ensureBlock();
 
-        // add the value declaration to the new loop body
-        if (declar) {
-          block.body.push(declar);
-        }
+        node.body.body.unshift(declar);
+
+        const nodes = builder.build({
+          CREATE_ITERATOR_HELPER: state.addHelper(builder.helper),
+          ITERATOR_HELPER: scope.generateUidIdentifier("iteratorHelper"),
+          STEP_KEY: t.identifier(stepKey),
+          OBJECT: node.right,
+          BODY: node.body,
+        });
+        const container = builder.getContainer(nodes);
 
-        // push the rest of the original loop body onto our new body
-        block.body = block.body.concat(node.body.body);
+        t.inherits(container[0], node);
+        t.inherits(container[0].body, node.body);
```

Comment:
Does `container` ever contain more than one element? If not, we can replace the `replaceWithMultiple` with `replaceWith` and drop the `parent.isLabeledStatement()` check.

Reply:
`container` always contains a single node. However:
1. We are doing `replaceWithMultiple(nodes)`, and  in non-loose mode `nodes` contains 2 nodes (the var declaration and the try statement)
2. We can't drop the `parent.isLabeledStatement()` check because in non-loose mode it the label must be moved to the inner loop. If we only did `path.replaceWith()` it would be attached to the `try` statement.

Reply:
>  If we only did path.replaceWith() it would be attached to the try statement.

I think that would actually be equivalent. The `finally` block should still fire correctly.

Reply:
But then, it doesn't work with `continue`.

Reply:
Ohhh. I was only thinking about `break`.